src/herder/HerderImpl.cpp

```diff
@@ -542,11 +543,12 @@ HerderImpl::emitEnvelope(SCPEnvelope const& envelope)
     broadcast(envelope);
 }
 
-TransactionQueue::AddResult
+std::pair<TransactionQueue::AddResult, TransactionResultPayloadPtr>
 HerderImpl::recvTransaction(TransactionFrameBasePtr tx, bool submittedFromSelf)
 {
     ZoneScoped;
-    TransactionQueue::AddResult result;
+    std::pair<TransactionQueue::AddResult, TransactionResultPayloadPtr> result{
+        TransactionQueue::AddResult::ADD_STATUS_COUNT, nullptr};
```

Comment:
Similar to `TransactionFrame`, the canonical user type for `TransactionResultPayload` is `TransactionResultPayloadPtr` (to make maintaining up to date state easier in the test class). Since this needs to be a pointer, I don't really see the point in wrapping an already verbose type in an optional when we have an underlying pointer type that is already nullable.

Reply:
the benefit of optional is that it conveys the intent correctly (a value may or may not be set) as opposed to shared pointer, which confusingly conveys shared ownership of the result object. I agree with Dima's comment elsewhere that letting test code dictate how we structure prod codepaths feels a bit wrong. That being said, it might be too challenging to change it at this time, given how tangled tx frame tests have become over the years.